ext/opcache/Optimizer/zend_inference.c

```diff
@@ -2085,8 +2089,43 @@ uint32_t zend_array_element_type(uint32_t t1, int write, int insert)
 	return tmp;
 }
 
+static inline uint32_t get_dim_array_key_type(uint32_t dim_type, zend_uchar dim_op_type, zval *dim_op) 
+{
+	uint32_t tmp = 0;
+
+	if (dim_op_type == IS_UNUSED) {
+		tmp |= MAY_BE_ARRAY_KEY_LONG;
+	} else {
+		if (dim_type & (MAY_BE_LONG|MAY_BE_FALSE|MAY_BE_TRUE|MAY_BE_RESOURCE|MAY_BE_DOUBLE)) {
+			tmp |= MAY_BE_ARRAY_KEY_LONG;
+		}
+		if (dim_type & MAY_BE_STRING) {
+			tmp |= MAY_BE_ARRAY_KEY_STRING;
+
+			if (dim_op_type != IS_CONST) {
+				tmp |= MAY_BE_ARRAY_KEY_LONG;
+			} else {
+				zend_ulong hval;
+				if (ZEND_HANDLE_NUMERIC(Z_STR_P(dim_op), hval)) {
+					tmp |= MAY_BE_ARRAY_KEY_LONG;
+				}
+			}
```

Comment:
It would be better to write this as:

```
		if (dim_type & MAY_BE_STRING) {
			if (dim_op_type != IS_CONST) {
				tmp |= MAY_BE_ARRAY_KEY_STRING | MAY_BE_ARRAY_KEY_LONG;
			} else {
				zend_ulong hval;
				if (ZEND_HANDLE_NUMERIC(Z_STR_P(dim_op), hval)) {
					tmp |= MAY_BE_ARRAY_KEY_LONG;
				} else {
					tmp |= MAY_BE_ARRAY_KEY_STRING;
				}
			}
		}
```

Otherwise we're adding MAY_BE_ARRAY_KEY_STRING even if we know numeric conversion will happen.

Reply:
Good call. 👍 Done in 45e6b24